plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftDeploymentCleaner.java

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.openshift.client;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Singleton;
+
+import org.eclipse.che.plugin.openshift.client.exception.OpenShiftException;
+import org.eclipse.che.plugin.openshift.client.kubernetes.KubernetesResourceUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodList;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.extensions.Deployment;
+import io.fabric8.kubernetes.api.model.extensions.DoneableDeployment;
+import io.fabric8.kubernetes.api.model.extensions.ReplicaSet;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.FilterWatchListDeletable;
+import io.fabric8.kubernetes.client.dsl.ScalableResource;
+import io.fabric8.openshift.api.model.Route;
+import io.fabric8.openshift.client.DefaultOpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftClient;
+
+@Singleton
+public class OpenShiftDeploymentCleaner {
+    private static final Logger LOG = LoggerFactory.getLogger(OpenShiftDeploymentCleaner.class);
+
+    public void cleanDeploymentResources(final String deploymentName, final String namespace) throws IOException {
+        scaleDownDeployment(deploymentName, namespace);
+        cleanUpWorkspaceResources(deploymentName, namespace);
+        waitUntilWorkspacePodIsDeleted(deploymentName, namespace);
+    }
+
+    private void scaleDownDeployment(String deploymentName, final String namespace) throws OpenShiftException {
+        try (OpenShiftClient openShiftClient = new DefaultOpenShiftClient()) {
+            ScalableResource<Deployment, DoneableDeployment> deployment = openShiftClient.extensions()
+                           .deployments()
+                           .inNamespace(namespace)
+                           .withName(deploymentName);
+
+            if (deployment != null) {
+                deployment.scale(0, true);
```

Comment:
@amisevsk btw in master I do not see that we use "terminationGracePeriodSeconds", So while testing it on minishift scale down does not happen fast and require some time - this might be the main reason of quota limit issue that was reported today